File: py/fls.py
from re import L
import numpy as np


import numpy as np

def update(mu, beta, P, R, x, y):
    # compute vector k (with old R and x)
    k = np.einsum('ij,j->j', R, x) / (1 + np.dot(x, np.einsum('ij,j->j', R, x)))
    # update P (with old R, k and x)
    new_P = R - np.einsum('kj,ji->ki', np.einsum('k,j->kj',k,x), R)
    # update R (with old P)
    new_R = P / mu
    e = y - np.dot(beta, x)
    new_beta = beta + e * k

    return new_beta, new_P, new_R


if __name__ == "__main__":
    factor = 0.95
    beta = np.zeros(4)
    P = np.eye(4)
    mu = (1-factor) / factor
    R = P / mu

    x = np.array([10., 12., 14., 3.])
    y = 12.0
    print()
    for i in range(480):
        beta, P, R = update(mu, beta, P, R, x, y)
    print(beta)

File: src/fls.c
/* Online Flexible least square
*/
#include <stdio.h>
#include "fls.h"

void fls_init(fls *ls, f32 factor)
{
    ls->factor = factor;
    v4_zero_(&(ls->beta));
    m4_eye_(&(ls->P));
    ls->mu = (1 - factor) / factor;
    ls->R = m4_sdiv(ls->mu, ls->P);
}

void fls_update (fls *ls, v4 x, f32 y)
{
    // compute vector k (with old R and x)
    v4 k = v4_sdiv(
        (1 + v4_dot(x, m4v4_mmul(ls->R, x))),
        m4v4_mmul(ls->R, x)
    );
    printf("k: \n");
    print4(k);
    // store old P (for updating R)
    m4 oldP;
    m4_copy(&(oldP), &(ls->P));
    // update P (with old R, k and x)
    ls->P = m4_msub(
        ls->R,
        m4_mmul(v4_mmul(k, x), ls->R)
    );
    // update R (with old P)
    ls->R = m4_sdiv(ls->mu, oldP);
    f32 e = y - v4_dot(ls->beta, x);
    ls->beta = v4_msum(ls->beta, v4_smul(e, k));
    print4(ls->beta);
    print4x4(ls->P.M);
    print4x4(ls->R.M);
}

f32  fls_predict(fls *ls, v4 x)
{
    return v4_dot(ls->beta, x);
}